Add support for local files

mre opened the pull request:
Fixes #21 

TODO:
* [x] Add support for base-url (`lychee deploy/specific/sub/page.html --base-url deploy/`) for absolute local links.
  Absolute local links (leading slash) require the base dir option to define the document root, which can be an external URL as well as a local directory path. So I was able to do liche -d build -r build/docs to check all local links, where the website is built into the build/docs directory and will be reached at https://example.org/docs finally. (see [comment](https://github.com/lycheeverse/lychee/issues/21#issuecomment-748467667))
* [x] Consider `--document-root` or `--base` defaults to the parent path of the file
* [x] Add flag to prevent network requests (similar to how `-x` prevents liche from checking external links)

MichaIng commented on README.md:
```diff
@@ -182,7 +182,8 @@ FLAGS:
 OPTIONS:
     -a, --accept <accept>                      Comma-separated list of accepted status codes for valid links
-    -b, --base-url <base-url>                  Base URL to check relative URLs
+    -b, --base <base>                          Base URL or website root directory to check relative URLs e.g.
```
I suggest to to call it "internal URLs" or even "absolute internal URLs" instead of "relative URLs". From my understanding:
- relative internal URL: `path/to/here` or `../path/to/there`
    - no leading slash, where a base path/URL can be derived from the input document's path/URL directly (appending + optionally resolving dot path elements), hence no additional option is required.
- absolute internal URL: `/path/to/elsewhere`
    - leading slash, where a base path/URL is indeed required, as it cannot be derived relative from the input document's path/URL.
- external URL: `https://example.org/stairway/to/heaven`

lebensterben replied:
base URL of HTML has a definition:

https://html.spec.whatwg.org/multipage/urls-and-fetching.html#document-base-url

Whatever option name is used, it must comply with the specification.

MichaIng replied:
Thanks for the links. So it's "path-relative-URL string" and "path-absolute-URL string" that I usually name internal links, while indeed they are both "relative-URL strings". So the "relative" in the first case means "relative path" while in the second case it means "relative to any leading component of the full URL" 😅.

A separation is still required, as for "path-relative-URL strings", no base path/URL is required, while for "path-absolute-URL strings" it is. And I'm not fully convinced that "Base URL or website root directory to check path-relative-URL strings e.g." is great here 🤔. On the other hand, reading it again, not to bad, and correct. Not me to decide 🙂.

mre commented:
Question: Should lychee check the existence of an `index.html` if a link points to a folder? E.g. if we find a link to `/about`, would we just check if the folder exists or if there is an actual HTML file that could be loaded?

I'm thinking we shouldn't check for the `index.html`, because the specifics heavily depend on the server configuration; e.g. Apache or NGINX might accept an `index.htm` or an `index.php` as well or show a directory listing if none of these don't exist. This behavior can be overwritten so we can't be sure unless we check the server config. That is a whole other rabbit hole, which we should probably avoid. It might be a caveat that we should document somewhere maybe. Thoughts?

MichaIng commented:
> I'm thinking we shouldn't check for the index

I agree for the same reasons. When the directory exists, then in 99% of cases something is served. Theoretically an `--index` option could be added for define accepted index files, but I cannot imagine that this would be much used.

mre commented on lychee-lib/src/collector.rs:
```diff
@@ -1,223 +1,10 @@
-    path::{Path, PathBuf},
-};
-
-use glob::glob_with;
```
Node: All of the removed type definitions below got moved into the the `types` module.

mre commented on lychee-lib/src/extract.rs:
```diff
@@ -1,82 +1,77 @@
+    Base, ErrorKind, Input, Request, Result, Uri,
+};
 
-impl<P: AsRef<Path>> From<P> for FileType {
```
Wasn't removed, just moved to `types/file.rs`

mre commented:
@lebensterben @MichaIng @pawroman:
This is ready for review from my side. Nevermind the publish check issues. That's because of changes to the lychee-lib, which aren't published yet. We hopefully find a solution for that in #305.

Apart from that, the tests are passing and I hope I covered all cases (let me know if I didn't!). Also if you can spare some time I'd be super thankful if you could try it on a local project like a static website for instance. Something like this should work:

```
cargo install --git https://github.com/lycheeverse/lychee --branch local-files
# Test local file links only
lychee --offline --base public public/index.html
# Test all the links
lychee --base public public/index.html
```

The `base` is needed if you have any non-relative links (e.g. `/about`).
This could be a little smarter by assuming the path to the root directory as base if it is not set (e.g. `public` in this case).
However I wasn't sure what to do in case of multiple inputs. We could use the base each input, but that would require to track the origin of a link throughout the run and I didn't want to add that functionality right away (YAGNI).

mre commented:
> Nevermind the publish check issues. That's because of changes to the lychee-lib, which aren't published yet. We hopefully find a solution for that in #305.

Well @dblock fixed that and the build is green now. Thanks! 😊

untitaker commented:
>Question: Should lychee check the existence of an index.html if a link points to a folder?

I think I disagree with the decision made here. I would suggest that you pick one or two static site hosters, such as gh pages and netlify, and document that your idea of a valid link matches those service's idea. Particularly gh pages offers no configuration as to what is considered an index page which imo already establishes strong conventions that most static site generators try to follow.

I think php files are a red herring because you'd have to execute them to find broken links, so for those usecases lychee is already a bad fit. I'd focus on pure static sites for the local fs usecases.

MichaIng commented:
@untitaker 
But all webservers support automatic directory indexes, so when a local link points to a directory, how do you know whether an actually contained `index.html` is required or whether it's aimed to be served as directory index? To not cause a lot of false positives by doing false assumptions via additional logic, you'd need to as well provide an option to define one or more valid index files, or whether required at all, which makes it more and more complicated. Mid- or long-term I'd agree to have such optionally, but would by default always succeed when the directory itself exist and the option is not specified.

For PHP files on the other hand, I don't see why it makes a difference, as the convention to use `index.php` is even stronger compared to index.html/xhtml/xml/htm/<empty> what what not that exist for static pages 😉. So no need to "execute" PHP (out of scope anyway) but simply check for the `index.php` index file. But again, IMHO this would be up to the user to define it as valid index via mentioned option. On dynamic websites, internal rewrites are more common, e.g. to have beautified URLs without `index.php` in the middle, which makes it harder to do everything right OOTB, but lychee has convenient options, now also with the exclude file, to handle such cases.

For now however, I think it is pretty great, @mre many thanks for your hard work the last days, I'll implement it into our workflows and show you a before/after.

MichaIng commented:
### Test on main website
#### lychee v0.7.1
```
Run ./lychee -X head -b 'http://127.0.0.1' --github-token '***' '**/*.md' '**/*.html'
📝 Summary
---------------------
🔍 Total..........299
✅ Successful.....299
⏳ Timeouts.........0
🔀 Redirected.......0
👻 Excluded.........0
🚫 Errors...........0
```
#### lychee v0.7.2 (local-files branch)
```
Run lychee -X head -b '.' --github-token '***' '**/*.md' '**/*.html'
📝 Summary
---------------------
🔍 Total..........637
✅ Successful.....625
⏳ Timeouts.........0
🔀 Redirected.......0
👻 Excluded.........0
🚫 Errors..........12

Errors in dietpi-software.html
✗ file:///home/runner/work/DietPi-Website/DietPi-Website/%23gettingstarted (Invalid file URI: file:///home/runner/work/DietPi-Website/DietPi-Website/%23gettingstarted)
✗ file:///home/runner/work/DietPi-Website/DietPi-Website/%23download (Invalid file URI: file:///home/runner/work/DietPi-Website/DietPi-Website/%23download)
✗ file:///home/runner/work/DietPi-Website/DietPi-Website/%23features (Invalid file URI: file:///home/runner/work/DietPi-Website/DietPi-Website/%23features)
✗ file:///home/runner/work/DietPi-Website/DietPi-Website/%23testimonials (Invalid file URI: file:///home/runner/work/DietPi-Website/DietPi-Website/%23testimonials)

Errors in contribute.html
✗ file:///home/runner/work/DietPi-Website/DietPi-Website/%23testimonials (Invalid file URI: file:///home/runner/work/DietPi-Website/DietPi-Website/%23testimonials)
✗ file:///home/runner/work/DietPi-Website/DietPi-Website/%23download (Invalid file URI: file:///home/runner/work/DietPi-Website/DietPi-Website/%23download)
✗ file:///home/runner/work/DietPi-Website/DietPi-Website/%23features (Invalid file URI: file:///home/runner/work/DietPi-Website/DietPi-Website/%23features)
✗ file:///home/runner/work/DietPi-Website/DietPi-Website/%23gettingstarted (Invalid file URI: file:///home/runner/work/DietPi-Website/DietPi-Website/%23gettingstarted)

Errors in stats.html
✗ file:///home/runner/work/DietPi-Website/DietPi-Website/%23gettingstarted (Invalid file URI: file:///home/runner/work/DietPi-Website/DietPi-Website/%23gettingstarted)
✗ file:///home/runner/work/DietPi-Website/DietPi-Website/%23features (Invalid file URI: file:///home/runner/work/DietPi-Website/DietPi-Website/%23features)
✗ file:///home/runner/work/DietPi-Website/DietPi-Website/%23download (Invalid file URI: file:///home/runner/work/DietPi-Website/DietPi-Website/%23download)
✗ file:///home/runner/work/DietPi-Website/DietPi-Website/%23testimonials (Invalid file URI: file:///home/runner/work/DietPi-Website/DietPi-Website/%23testimonials)
```
It finds more than doubled amount of links, and, looking at the failures, I guess it interprets page anchors, like `/#gettingstarted`, to link to the `id="gettingstarted"` element on the main index page, as part of the path, which then becomes invalid of course, and it explains the large amount of additional links, for every unique page anchor one.

mre commented:
So if I understand correctly it should strip off those anchors right? 😅

MichaIng commented:
> So if I understand correctly it should strip off those anchors right? 😅

Theoretically, if the URL-encoding done somewhere would be removed, the "fragment" (now I remember the correct word for this 😄) may be ignored, stripped and handled as one URL already, as this is true for external URLs with current lychee releases. But if indeed local URLs are handled differently in the first place, so that not (only) the URL-encoding is the issue, then yes.

mre commented:
> I would suggest that you pick one or two static site hosters, such as gh pages and netlify, and document that your idea of a valid link matches those service's idea. Particularly gh pages offers no configuration as to what is considered an index page which imo already establishes strong conventions that most static site generators try to follow.

Yeah that's a tricky one. My fear is that whatever standard we'll pick will lead to weird inconsistencies down the road. I agree that Github Pages is sort of a quasi-standard at the moment, but I wouldn't bet on _any_ current solution for the future. I'm just trying to be conservative with my guarantees here so that there are no big surprises.

If anything, we should think about making this configurable. We could add a param like `--index_page index.html`, which would accept a regex and would make a index files mandatory for each directory. This way users could change the behavior to match their own static website layout.

untitaker commented:
I think it's more dangerous and less conservative to over-accept links as valid rather than under-accept. If you choose to opt for over-accepting to avoid floods of errors by default, I fear that the experience over time will be that newer versions lychee will find less and less 404s in its default mode as support for more usecases is added

>For PHP files on the other hand, I don't see why it makes a difference, as the convention to use index.php is even stronger compared to index.html/xhtml/xml/htm/ what what not that exist for static pages wink. So no need to "execute" PHP (out of scope anyway) but simply check for the index.php index file.

Yes, you can validate a link to index.php without parsing the PHP file. But I would argue that in such a hypothetical usecase lychee is already a bad fit anyway because one would likely want to check the entire website for broken links, not just individual files that lychee happens to be able to parse (i.e. they will likely opt for HTTP-based link checking). My broader argument here is that lychee-on-fs is only realistically usable on fully static sites so why even think about PHP?

untitaker commented:
bug: If I run `lychee --offline -b . '**/*.html'` on [sentry-docs](https://github.com/getsentry/sentry-docs), I get:

```
Error: Failed to read file: `page-data/404.html`, reason: Is a directory (os error 21)
```

The reason for this is that the glob matches a directory named 404.html (don't ask me what that is, it's some random stuff Gatsby generated), and that cannot be read as a file. Would argue that the glob should never match dirs, but not sure what arguments I'm passing.

So I resorted to:

```
~/projects/lychee/target/release/lychee --offline -b . $(rg --files -thtml)
```

There's some overhead imposed by `rg --files`, and then it takes 3-4 minutes to run. I then hit the same problem as @MichaIng with anchors not resolving to filepaths.

MichaIng commented:
> Yes, you can validate a link to index.php without parsing the PHP file. But I would argue that in such a hypothetical usecase lychee is already a bad fit anyway because one would likely want to check the entire website for broken links

Yes that is true, and I think it is out of scope to hassle with this. For a dynamic website one can spin up a webserver, as well in CI, and then use `lychee` to check the served website. #165 will make this even easier.

> The reason for this is that the glob matches a directory named 404.html

Lol, what the hell creates a directory named like that 😄. However, I think directories should be skipped silently and not throw an error, at least when being matched by a glob instead of given as explicit input (if it is sufficiently easy to implement skipping for globs only).

untitaker commented:
@mre this branch does not build for me, need to:

```diff
diff --git a/lychee-lib/src/client.rs b/lychee-lib/src/client.rs
index e8d477f..ed49213 100644
--- a/lychee-lib/src/client.rs
+++ b/lychee-lib/src/client.rs
@@ -185,7 +185,7 @@ impl Client {
             match self.check_website(&uri).await {
                 Status::Ok(code) if self.require_https && uri.scheme() == "http" => {
                     let mut https_uri = uri.clone();
-                    https_uri.url.set_scheme("https").unwrap();
+                    https_uri.inner.set_scheme("https").unwrap();
                     if self.check_website(&https_uri).await.is_success() {
                         Status::Error(Box::new(ErrorKind::InsecureURL(https_uri)))
                     } else {

```

mre commented:
> Error: Failed to read file: `page-data/404.html`, reason: Is a directory (os error 21)

That's fixed now.

> @mre this branch does not build for me, need to:

Fixed; messed up the rebase. Sorry.

> I then hit the same problem as @MichaIng with anchors not resolving to filepaths.

Fixed.

> I think it's more dangerous and less conservative to over-accept links as valid rather than under-accept.

My experience with other link-checkers so far is that they have a high false-positive rate, which makes checking tedious. I'd rather have a tool that reliably finds broken links instead of one "crying wolf" once too often. 😉

mre commented:
@MichaIng, @untitaker can you give this another spin? 🙇 I'm looking for correctness rather than performance here at the moment. We'll have to do a bigger refactor to fix the perf issues once we have basic file support.

MichaIng commented on lychee-lib/src/helpers/url.rs:
```diff
@@ -0,0 +1,85 @@
+        Some((path_without_params, _params)) => path_without_params,
+        None => path,
+    };
+    path
```
It would be smart to do the query string split first, so that the dedicated fragment split can be skipped in case. Would something like this work?
```rust
    let path = match path.split_once('?') {
        Some((path_without_params, _params)) => path_without_params,
        None => match url.split_once('#') {
            Some((path_without_fragment, _fragment)) => path_without_fragment,
            None => url,
        };
    };
```
- So if there is a query string, strip and return, as the fragment has been stripped off together with the query string.
- Only if there is no query string, strip off the fragment.

mre replied:
Makes sense. I changed that now. Thanks!

untitaker replied:
What about questionmarks in anchors/fragments?

MichaIng replied:
> What about questionmarks in anchors/fragments?

Indeed I also though about that, see my edit. I think it is very care, but not worth it cause false negatives for such a little performance benefit. But as noted, best is probably to remove fragments from all URLs (as long as they are not verified) and e.g. add them in a different way to the output to not give a false impression that they would be verified: https://github.com/lycheeverse/lychee/pull/262#issuecomment-914297175

mre replied:
Reverted the change and added a test. It never crossed my mind that fragments could contain question marks. Wow.

MichaIng commented:
Goes through now: https://github.com/MichaIng/DietPi-Website/runs/3532854487
I wonder why there are still much more links detected then before. Is there a reason why the verbosity flag has no effect if lychee is not attached to a regular terminal?

mre commented:
> Is there a reason why the verbosity flag has no effect if lychee is not attached to a regular terminal?

Can you try adding `--no-progress`?

MichaIng commented:
Ah, relative links links `<a href="path/to/other.html">` were not supported at all before, but only absolute paths (when base URL given), like `<a href="/path/to/other.html">`, right?

Running with `--no-progress`: https://github.com/MichaIng/DietPi-Website/runs/3532983710

Another thing, indirectly related since fragments are now removed for local files:
```
✔ https://dietpi.com/#download [200 OK]
✔ https://dietpi.com/#features [200 OK]
```
Since fragments are currently not checked, wouldn't it be reasonable to remove them in the first place from all URIs, so that the above two are handled as one?

mre commented:
Nice work. Thanks for the feedback. 😃 

About fragments: there is remove for improvement.
We can't know if the fragments in this case are valid, but we show them as OK:

```
✔ https://dietpi.com/#download [200 OK]
✔ https://dietpi.com/#features [200 OK]
```

My initial reaction was "let's fix fragments!" but that will be harder, given that we'd have to parse the file to see if it exists. And that file can be anything from `index.html` over `index.php` to `default.aspx`. We just can't know without some webserver or at least some config. 

Removing the fragments would reduce the workload and trim the output, but it wouldn't be fully correct. After all, the fragments serve a purpose and checking deep-links is an important feature that we should cover in the end. 

The other option is to detect and show fragments, but improve the reporting.
Coming back to my original idea, we could have an argument to define the index page like `--index index.html` or `--index (index.htm|index.html)` (which could be the default) and if we encounter an anchor then...

* if `/path/index.html` exists, check if the anchor is there by parsing the file
* if `/path/index.html` does not exist and `--index` was not t set, show it as unknown
* if `/path/index.html` does not exist and `--index` was set, show it as error

What do you think of that?

MichaIng commented:
On fragments:
- Checking those must definitely stay an optional non-default feature, as scanning the whole HTML document increases traffic significantly, does not work with `-X head`, works only with actual HTML responses etc. But yes, having this option would be nice.
- As long as fragments are not actually checked, I'm not sure whether printing them as part of the checked URLs is correct or gives a false impression. On the one hand, users may miss the fragments (when not knowing/thinking about the topic) in output, on the other hand they may get the impression that those are checked, when having them in the output.
- So yes, showing them but enhancing the output may be a good compromise, like, splitting it away from the URL, before processing it (so that it also benefits from URI cache, having less URLs actually checked), but keeping it in an extra variable. Then printing it in parenthesis behind the URL, like:
    ```
    ✔ https://dietpi.com/ (fragment: #download) [200 OK]
    ```
    Adding a whole sentence is too much, but like that there is somehow an indication that the fragment is handled differently?

Not sure how fragment checks and local index file checks are related? I mean of course one may also want to validate fragments in local files, and in this case no full HTML download is required, so that fragment/ID validation is possible with less overhead, but its still two generally independent features, isn't it?

I'd still by default accept directories without `index.html`, but having an "unknown" status may be a compromise. The default `index.html` should be okay nowadays, I think `index.htm` isn't used widely anymore and also isn't a default on current Apache2, Nginx or Lighttpd. But if it's still wanted, the use `--index 'index.htm(|l)'` 😉.

MichaIng commented:
Amazing! I just tested it on our documentation repository:
- Without local file support, we needed to run a local webserver to check internal links, using the MkDocs internal one. To have internal links from files in each sub directory correctly attached to the correct base URL, we needed to loop through every directory and call lychee on each with the directory used path for the base URL. Actually strange that this was necessary, since it seems like relative internal paths (no leading `/`) were not checked before? Maybe also something changed about that with v0.7.1 or some releases before? However, this, and the fact that all internal links caused an outgoing + incoming traffic, made it highly inefficient, so that it took 4 - 10 minutes:
    - Workflow: https://github.com/MichaIng/DietPi-Docs/blob/455b854/.github/workflows/actions.yml#L67
    - Last log: https://github.com/MichaIng/DietPi-Docs/runs/3536642975
- Now that internal relative paths are resolved correctly, we can check everything with a single call, taking awesome 10 seconds (!):
    - Workflow: https://github.com/MichaIng/DietPi-Docs/blob/6553a3e/.github/workflows/actions.yml#L54
    - Last log: https://github.com/MichaIng/DietPi-Docs/runs/3536747646
 
One thing now I recognised:
- I counted all links of all individual lychee calls with old workflow together and get: `4061`
- With the new workflow I get: `3326`
- If I understand it right, even if repeated URLs across files are checked only once, they are still counted once for each appearance in each file in the summary, right? Then the reason for the decreased URL count cannot be due to repeated URLs now being checked in one call?
- But how is it now for internal/local file checks, especially since now fragments and query strings are stripped, are `path/to?some#fragment` and `path/to?other#fragment` and `path/to`, when collected in the same file, now counted as one URL in the summary or as three?

mre commented:
> But how is it now for internal/local file checks, especially since now fragments and query strings are stripped, are path/to?some#fragment and path/to?other#fragment and path/to, when collected in the same file, now counted as one URL in the summary or as three?

**For remote files we keep the fragments and params as a web-server can handle them.**

```
> cat input.md                                                                                                                               ✘
[fragment](path/to?some#fragment)
[bar](path/to?other#fragment)
[baz](path/to)
```

```
> lychee --base https://example.org/ input.md                                                    ✘
📝 Summary
---------------------
🔍 Total............3
✅ Successful.......0
⏳ Timeouts.........0
🔀 Redirected.......0
👻 Excluded.........0
🚫 Errors...........3

Errors in input.md
✗ https://example.org/path/to?other#fragment (HTTP status client error (404 Not Found) for url (https://example.org/path/to?other#fragment))
✗ https://example.org/path/to?some#fragment (HTTP status client error (404 Not Found) for url (https://example.org/path/to?some#fragment))
✗ https://example.org/path/to (HTTP status client error (404 Not Found) for url (https://example.org/path/to))
```
**For local files it will count as one (since as you said fragments and query params get stripped).**

```
> mkdir -p path/to
```

```
> lychee --offline --base . input.md
📝 Summary
---------------------
🔍 Total............1
✅ Successful.......1
⏳ Timeouts.........0
🔀 Redirected.......0
👻 Excluded.........0
🚫 Errors...........0
```

MichaIng commented:
> For local files it will count as one (since as you said fragments and query params get stripped).

Okay, that explains it all.

> For remote files we keep the fragments and params as a web-server can handle them.

Jep, removing them, or at least ignoring them when checking the URI cache for an existing result, would only be a performance enhancement, respectively for reduced requests. I just recognised that we get a high amount of `429 Too Many Requests` from GitHub, despite a given token, especially on one page with a very high amount of GitHub links 😄.

mre commented:
>  just recognised that we get a high amount of 429 Too Many Requests from GitHub, despite a given token, especially on one page with a very high amount of GitHub links 😄.

It would be nice to add some exponential backoff and request delays eventually. See #36.

mre commented:
From 4 - 10 minutes to 10 seconds is a big win (60x best case perf improvement, hehe). Of course it's network I/O vs file I/O, but it's an important use-case for many people. 😊

mre commented:
There is still a lot to do, but from my side this is ready to merge now. All the rest (perf issues with progress-bar, fragment handling, I/O issues with glob) should be moved to a separate issue. Any objections?

MichaIng commented:
Agreed. One last thing I wanted to test is whether special characters in the base path or from root to base are URL-encoded and do then fail the local file check because of this. I can do that tomorrow, but if someone is faster, it can be merged from my end.

mre commented:
This _should_ be fine, but please double-check.

```
> lychee --offline --base '@$%^&&*()-foo-- bar' '@$%^&&*()-foo-- bar/index.html' --verbose
✔ file:///private/tmp/test/@$%25^&&*()-foo--%20bar/about [200 OK]
✔ file:///private/tmp/test/@$%25^&&*()-foo--%20bar [200 OK]

📝 Summary
---------------------
🔍 Total............2
✅ Successful.......2
⏳ Timeouts.........0
🔀 Redirected.......0
👻 Excluded.........0
🚫 Errors...........0
````

untitaker commented:
I haven't gotten around to checking yesterday but from my last test it seemed to me that lychee discovered slightly fewer links in sentry-docs than hyperlink. I haven't had time to investigate this and it doesn't help that hyperlink has very little debug output.

mre commented:
Had a look at hyperlink yesterday and I'm impressed by the readability and clarity of the code without sacrificing performance. Good job there! Also if you find anything regarding the missing links I'd be super interested.

MichaIng commented:
@untitaker 
I guess the missing links are for the same reason explained above: https://github.com/lycheeverse/lychee/pull/262#issuecomment-914661965
Some verbose output of `hyperlink` could indeed verify it, when internal links with only different fragments (or query strings) are counted multiple times.

Looks like for the local file check parts, one can learn from [hyperlink](https://github.com/untitaker/hyperlink) then, at least for some aspects where the external/HTTP check aspects do not interfere. But step-by-step, for now I'm extremely happy that local files are supported at all and we get such an enormous speedup by this 😃.

> This should be fine, but please double-check.

I'll do another test, but looks good indeed as the `file://` URI check finally follows the same rules and logic regarding URL encoding, and get's decoded again in [to_file_path](https://docs.rs/url/2.2.2/url/struct.Url.html#method.to_file_path). So the only issue was the [from_file_path](https://docs.rs/url/2.2.2/url/struct.Url.html#method.from_file_path) URL-encoding the `#`, which is not subject of any transferred URL and hence must not be URL-encoded but skipped and interpreted only by the client itself after receiving the HTML response. To avoid fase decoding, e.g. a literal `+` in the file path into a space, it is even required to encode the base path/root, as long as `to_file_path` is used on it. Last thing to verify is how it behaves on a mixture, e.g. URL-encoded internal link, to assure it does not get double-encoded and hence stays encoded when checking the file path.

MichaIng commented:
Okay, it is still an issue if the URL given in the input file is URL-encoded already, as it indeed gets double-encoded:
```
# cat README.md
<a href="/web site/">
<a href="/web%20site">

# lychee -vb .
✔ file:///root/web%20site [200 OK]
✗ file:///root/web%2520site (Invalid file URI: file:///root/web%2520site)

📝 Summary
---------------------
🔍 Total............2
✅ Successful.......1
⏳ Timeouts.........0
🔀 Redirected.......0
👻 Excluded.........0
🚫 Errors...........1

Errors in README.md
✗ file:///root/web%2520site (Invalid file URI: file:///root/web%2520site)

# curl -I 'file:///root/web%20site'
Content-Length: 4096
Accept-ranges: bytes
Last-Modified: Wed, 08 Sep 2021 13:22:34 GMT
```

For the base/root path it is however required:
```
~/web site# lychee -vb .
✔ file:///root/web%20site/web%20site [200 OK]
✗ file:///root/web%20site/web%2520site (Invalid file URI: file:///root/web%20site/web%2520site)

📝 Summary
---------------------
🔍 Total............2
✅ Successful.......1
⏳ Timeouts.........0
🔀 Redirected.......0
👻 Excluded.........0
🚫 Errors...........1

Errors in README.md
✗ file:///root/web%20site/web%2520site (Invalid file URI: file:///root/web%20site/web%2520site)

~/web site# cd ..
~# lychee -vb 'web site'
✔ file:///root/web%20site/web%20site [200 OK]
✗ file:///root/web%20site/web%2520site (Invalid file URI: file:///root/web%20site/web%2520site)

📝 Summary
---------------------
🔍 Total............2
✅ Successful.......1
⏳ Timeouts.........0
🔀 Redirected.......0
👻 Excluded.........0
🚫 Errors...........1

Errors in README.md
✗ file:///root/web%20site/web%2520site (Invalid file URI: file:///root/web%20site/web%2520site)
```

So to do it correctly, I'm afraid `from_file_path` must only be used on root/base path, but not on the part given in the link, which are expected to be encoded already (else it is correct to handle it as error), but this needs to be instead appended literally. It is an edge case, especially since the query string is stripped and a website directory structure will rarely have any special characters or space in directory or file names, but it is possible.

mre commented:
Another nice catch. This should be resolved now as well.
I had to look into how the url crate handles encoding. Turns out they encode the URL automatically and there is no way to disable that functionality. Since splitting the URL manually and encoding the individual parts would sort of defeat the point of using the url crate in the first place, I decided to do the simple thing and percent_decode the path section before passing it to `Url::from_file_path`. Added a test for that but let me know if that doesn't make any sense. 
Other than that, +1?

mre commented:
I've added a test for that case:

```rust
#[test]
fn test_create_uri_from_path() {
    let result = create_uri_from_path(&PathBuf::from("/README.md"), &None, "test+encoding").unwrap();
    assert_eq!(result.as_str(), "file:///test+encoding");
}
```

Is that what you meant?

MichaIng commented:
I just realised that indeed only the path within the link is decoded, not the full path, so the theoretical case of a literal percent code in the root/base path shouldn't cause an issue. Made some tests:
```
~# ls web+site
 README.md 'web site'

~# cat web+site/README.md
<a href="/web site/">
<a href="/web%20site">

~# lychee -vb 'web+site' 'web+site/README.md'
✔ file:///root/web+site/web%20site [200 OK]

~/web+site# lychee -vb .
✔ file:///root/web+site/web%20site [200 OK]
```
- So in both cases, with `+` in the given base path or in the root path, it is taken literally and hence succeeds the check 👍.
- I actually would have expected that the `+` itself is URL encoded by `from_file_path` to `%2B`, like the space in the sub directory correctly is to `%20` 🤔. **EDIT: All correct, see further below!**
- While double decoding of the link path is now not an issue anymore, un-encoded characters in the link do succeed. The `%20` is decoded into space, so the two links are equal, then `from_file_path` encodes both to `%20`. Most browsers convert encode spaces (and probably other special characters) automatically, but not all clients, so strictly seen this is a false positive:
    ```
    # curl -f 'https://dietpi.com/test%20dir/'
    # curl -f 'https://dietpi.com/test dir/'
    curl: (22) The requested URL returned error: 400
    ```
With a real %-code:
```
~# lychee -vb 'web%20site' 'web%20site/README.md'
✔ file:///root/web%2520site/web%20site [200 OK]

~/web%20site# lychee -vb .
✔ file:///root/web%2520site/web%20site [200 OK]
```
- So the literal `%` character, as part of either the base path or the root path, is correctly skipped for `percent_decode_str` but then encoded to `%25` by `from_file_path` 👍.

Another one:
```
~/web+site# cat index.html
<a href="/web%2Bsite/">

~/web+site# ls
index.html  web+site

~/web+site# lychee -vb . 'index.html'
✔ file:///root/web+site/web+site [200 OK]
```
- So the `%2B` is decoded correctly to `+`, but then not encoded anymore by `from_file_path`. Makes sense as it was the same with the `+` in the base path. Looks like at least these functions do neither see the `+` character as something that needs to be decoded (to space), nor as something that needs to be URL-encoded. I just tested it in the browser, and while it automatically changes a space in the address bar into `%20`, it leaves the `+` untouched. Further checking on this: https://stackoverflow.com/a/1006074/16145737
    So if I understand it correctly, the `+` has indeed officially no special meaning in URL paths, but only in the query string part, interpreted as space by PHP, and otherwise by some RFC violating behaviour of .NET or so. Since we strip the query string, it is hence perfectly fine that the `+` is neither decoded, nor encoded but just left untouched.

Last test:
```
~/web+site# ls
index.html  web%20site

~/web+site# cat index.html
<a href="/web%20site/">

~/web+site# lychee -vb . 'index.html'
✗ file:///root/web+site/web%20site (Invalid file URI: file:///root/web+site/web%20site)

~/web+site# cat index.html
<a href="/web%2520site/">
~/web+site# lychee -vb . 'index.html'

```
- In this case, where the link path contains a literally meant URL code, it is required to have it encoded in the link.
- This is now inconsistent with the fact that a space or other special characters do not need to be encoded in the link.
_____________
So as a result, the only left issue is that special characters, which are not co-incidentally part of an URL coding, do not need to be URL encoded in the input file/link. While most browsers handle un-encoded spaces and probably other characters gracefully, it is at least bad practice and fails definitely `curl` and some other command line tools and libraries. If there is co-incidentally a literally meant URL code in the link path, it is decoded by `percent_decode_str` and hence in this case needs to be URL encoded in the input file/link, which is an inconsistency, though one that becomes apparent in a highly unlikely case.

IMHO it is good enough to go like that. Probably a note could be left in the code, so that in case of a refactoring, probably this left issue and inconsistency can be resolved. As said, for this the base and root path elements need to be URL-encoded, while the input link path element needs to be appended untouched, so that it is correctly expected to be URL encoded already.

mre commented:
Good point. I added a TODO so that we don't forget.

mre commented:
Merged! 🎉 
Thanks for your help @MichaIng and @untitaker.

untitaker commented:
@MichaIng re discrepancy in link count: will file follow-up ticket if I find anything, in principle anchor checking is disabled in hyperlink by default

untitaker commented:
Great work, sorry I didn't really catch up with the last couple messages at all! I'll do some more functional testing at a later point but I think yall got it covered...

mre commented:
It wouldn't have worked without your inspiration, so thanks for that. I learned a lot.

untitaker commented:
@mre 

>Had a look at hyperlink yesterday and I'm impressed by the readability and clarity of the code without sacrificing performance. Good job there! Also if you find anything regarding the missing links I'd be super interested.

Thanks! There's def a lack of code comments and a lot of premature optimization, though the latter is sort of the selling point... But yeah it's been interesting with regard to performance for sure
